rfc/reorder-waiting-executions.md

```diff
@@ -0,0 +1,116 @@
+| **SIG / WG**   | _Applicable SIG(s) or Working Group(s)_ |
+
+
+## Overview
+
+_Currently the waiting executions of a pipeline runs in FIFO order. But due to certain project urgencies and release purpose, we quite often need to change the executions order of a pipeline._
+
+_Implementing the Re-order functionality helps to run the waiting executions in the order the user wants_
+
+### Goals and Non-Goals
+
+Goals:
+
+_Introduce an option through Spin CLI and Spinnaker dashboard(GUI) to move the specific waiting execution of a Pipeline towards UP/DOWN or Top/Bottom from the current position, which is set in FIFO order by default_
+
+_What problems are not not trying to solve?_
+
+## Motivation and Rationale
+
+_Currently it is not possible If the user wants to prioritize specific execution of a pipeline. Adding Re-order feature from CLI/GUI will help end-users and operators to change the priority order of a waiting execution for release purpose or any project urgencies._
+
+
+## Timeline
+_The code for CLI command reorder implementation is almost ready for review. 
+Planning to implement this feature in two Phases._
+
+_Phase-1: completing CLI command implementation and make this feature available for customers to use and work on feedback if any._
+
+_Phase-2: Implementing GUI for reorder functionality by using the backend that is available from CLI reorder implementation in Phase-1._
+
+_Proposed timeline for Phase-1 would be 1-2 weeks as this feature is demoed for some Spinnaker users and got the feedback.
+Phase-2 would be 3-4 weeks as there are some existing issues in the waiting executions area of Spinnaker dashboard that needs to be fixed._
+
+## Design
+
+_The Re-ordering waiting executions of a pipeline can be done through Spin CLI command and Spin dashboard._
+
+_Introducing a new CLI command for reorder functionality of a waiting executions, the command will be used to Re-order waiting execution to UP/DOWN or Top/Bottom based on the current position of the provided execution._
+
+_Example Spin CLI Command_
+ >>spin pipeline execution reorder --execution-id 01GJQ601MC3SBACVNZB0XC5GHJ --reorder-action UP
+
+_New buttons will be added in Spinnaker dashboard(GUI) for each waiting execution to move the specific execution towards UP/Down or Top/Bottom based on the current position of the execution._
+
+_New Gate API endpoint will be created for external access_
+>>:8084/pipelines/<execution_id>/reorder?reorderAction=<UP/Down/Top/Bottom>
+
+_New Orca API endpoint will be created to access the same way_
+>>:8083/pipelines/<execution_id>/reorder?reorderAction=<UP/Down/Top/Bottom>
+
+_The main business logic of the feature lies in Spin Orca service, where the **CompoundExecutionOperator** will update the Redis repo using the existing class **RedisExecutionRepository** and update the pending executions to the user selected position in the Redis repo using the Class **RedisPendingExecutionService**._
```

Comment:
Note, would need to handle SQL as well if this moved forward.

Reply:
Thank you for reviewing this, Jason. Do we need this to be implemented for DualPendingExecutionService and InMemoryPendingExecutionService too.